src/inventorymanager.cpp

```diff
@@ -154,94 +170,257 @@ IMoveAction::IMoveAction(std::istream &is, bool somewhere) :
+	else if (ma.to_inv.type == InventoryLocation::NODEMETA)
+		sa->nodemeta_inventory_OnPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::PLAYER)
+		sa->player_inventory_OnPut(ma, item, player);
 	else
 		assert(false);
 
-	if (from_inv.type == InventoryLocation::DETACHED)
-		sa->detached_inventory_OnTake(*this, src_item, player);
-	else if (from_inv.type == InventoryLocation::NODEMETA)
-		sa->nodemeta_inventory_OnTake(*this, src_item, player);
-	else if (from_inv.type == InventoryLocation::PLAYER)
-		sa->player_inventory_OnTake(*this, src_item, player);
+	if (ma.from_inv.type == InventoryLocation::DETACHED)
+		sa->detached_inventory_OnTake(ma, item, player);
+	else if (ma.from_inv.type == InventoryLocation::NODEMETA)
+		sa->nodemeta_inventory_OnTake(ma, item, player);
+	else if (ma.from_inv.type == InventoryLocation::PLAYER)
+		sa->player_inventory_OnTake(ma, item, player);
 	else
 		assert(false);
 }
 
-void IMoveAction::onMove(int count, ServerActiveObject *player) const
+static void onMove(const MoveAction& ma, int count, ServerActiveObject *player)
 {
 	ServerScripting *sa = PLAYER_TO_SA(player);
-	if (from_inv.type == InventoryLocation::DETACHED)
-		sa->detached_inventory_OnMove(*this, count, player);
-	else if (from_inv.type == InventoryLocation::NODEMETA)
-		sa->nodemeta_inventory_OnMove(*this, count, player);
-	else if (from_inv.type == InventoryLocation::PLAYER)
-		sa->player_inventory_OnMove(*this, count, player);
+	if (ma.from_inv.type == InventoryLocation::DETACHED)
+		sa->detached_inventory_OnMove(ma, count, player);
+	else if (ma.from_inv.type == InventoryLocation::NODEMETA)
+		sa->nodemeta_inventory_OnMove(ma, count, player);
+	else if (ma.from_inv.type == InventoryLocation::PLAYER)
+		sa->player_inventory_OnMove(ma, count, player);
 	else
 		assert(false);
 }
 
-int IMoveAction::allowPut(const ItemStack &dst_item, ServerActiveObject *player) const
+static int allowPut(const MoveAction& ma, const ItemStack &item, ServerActiveObject *player)
 {
 	ServerScripting *sa = PLAYER_TO_SA(player);
-	int dst_can_put_count = 0xffff;
-	if (to_inv.type == InventoryLocation::DETACHED)
-		dst_can_put_count = sa->detached_inventory_AllowPut(*this, dst_item, player);
-	else if (to_inv.type == InventoryLocation::NODEMETA)
-		dst_can_put_count = sa->nodemeta_inventory_AllowPut(*this, dst_item, player);
-	else if (to_inv.type == InventoryLocation::PLAYER)
-		dst_can_put_count = sa->player_inventory_AllowPut(*this, dst_item, player);
+	int can_put_count = 0xffff;
+	if (ma.to_inv.type == InventoryLocation::DETACHED)
+		can_put_count = sa->detached_inventory_AllowPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::NODEMETA)
+		can_put_count = sa->nodemeta_inventory_AllowPut(ma, item, player);
+	else if (ma.to_inv.type == InventoryLocation::PLAYER)
+		can_put_count = sa->player_inventory_AllowPut(ma, item, player);
 	else
 		assert(false);
-	return dst_can_put_count;
+	return can_put_count;
```

Comment:
would be a good idea imo
or is -1 a valid value too?
```suggestion
	assert(can_put_count >= 0 && can_put_count <= 0xffff);
	return can_put_count;
```

Reply:
-1 is a valid value too, https://github.com/minetest/minetest/blob/8fded9d990d1e1142b630b24c96f1a2c959fef36/doc/lua_api.txt#L8640

Reply:
Should we check that it's < 65535 then?

Reply:
What is your goal here? It makes little sense for a mod to return a count > 65535, but it doesn't break anything. The code still would work as expected.

Reply:
The intent here is to check for invalid situations early, preventing potential problems.

Reply:
Okay, I don't mind adding it